clients/src/test/java/org/apache/kafka/common/metrics/OtlpMetricsReporter.java

```diff
@@ -0,0 +1,190 @@
+                return;
+            }
+
+            MetricsData metricsData = ClientTelemetryUtils.deserializeMetricsData(payload.data());
+            if (metricsData == null || metricsData.getResourceMetricsCount() == 0) {
+                log.warn("exportMetrics - No metrics available, suspended export");
+                return;
+            }
+
+            List<ResourceMetrics> metrics = metricsData.getResourceMetricsList();
+
+            // enhance metrics with Broker-added labels
+            if (appendClientLabels) {
+                Map<String, String> labels = getClientLabels(context, payload);
+                metrics = enhanceMetrics(metrics, labels);
+            }
+
+            grpcService.export(metrics);
+        };
+    }
+
+
+    /**
+     * Returns a list of ResourceMetrics where each metrics is enhanced by adding client information as
+     * additional resource level attributes
+     * @param resourceMetrics resource metrics sent by client
+     * @param labels broker added labels containing client information
+     * @return enhanced ResourceMetrics list
+     */
+    public List<ResourceMetrics> enhanceMetrics(List<ResourceMetrics> resourceMetrics,
+                                                Map<String, String> labels) {
+
+        List<ResourceMetrics> updatedResourceMetrics = new ArrayList<>();
+
+        resourceMetrics.forEach(rm -> {
+            Resource.Builder resource = rm.getResource().toBuilder();
+
+            labels.forEach((k, v) -> resource.addAttributes(
+                    KeyValue.newBuilder()
+                            .setKey(k)
+                            .setValue(AnyValue.newBuilder().setStringValue(v))
+                            .build()
+                    ));
+
+            ResourceMetrics updatedMetric = rm.toBuilder()
+                    .setResource(resource.build())
+                    .build();
+
+            updatedResourceMetrics.add(updatedMetric);
+        });
+
+        return updatedResourceMetrics;
+    }
+
+    public static class ClientLabels {
+        public static final String RESOURCE_LABEL_PREFIX = "";
+        public static final String CLIENT_ID = RESOURCE_LABEL_PREFIX + "client.id";
+        public static final String CLIENT_INSTANCE_ID = RESOURCE_LABEL_PREFIX + "client.instance.id";
+        public static final String CLIENT_SOFTWARE_NAME = RESOURCE_LABEL_PREFIX + "client.software.name";
+        public static final String CLIENT_SOFTWARE_VERSION = RESOURCE_LABEL_PREFIX + "client.software.version";
+        public static final String CLIENT_SOURCE_ADDRESS = RESOURCE_LABEL_PREFIX + "client.source.address";
+        public static final String PRINCIPAL = RESOURCE_LABEL_PREFIX + "principal";
+    }
+
+    public Map<String, String> getClientLabels(AuthorizableRequestContext context, ClientTelemetryPayload payload) {
+        RequestContext requestContext = (RequestContext) context;
```

Comment:
Is it possible for any of the values that are inserted into the the `labels` map to be `null`? Is it OK if they are `null`?

Reply:
Updated this section